tools/src/main/scala/scala/scalanative/optimizer/pass/InstCombine.scala

```diff
@@ -0,0 +1,228 @@
+
+          case _ => inst
+        }
+
+      case Let(n, Op.Select(cond, thenv, elsev)) =>
+        defop(cond) match {
+          // select (c xor true) then a else b  =  select c then b else a
+          case Some(Op.Bin(Xor, _, negCond, Val.True)) =>
+            Let(n, Op.Select(negCond, elsev, thenv))
+
+          case _ => inst
+        }
+
+      case If(cond, thenp, elsep) =>
+        defop(cond) match {
+          // if (c xor true) then a else b  =  if c then b else a
+          case Some(Op.Bin(Xor, _, negCond, Val.True)) =>
+            If(negCond, elsep, thenp)
+
+          case _ => inst
+        }
+
+      case Let(n, op @ Op.Bin(_, _, lhs, rhs)) =>
+        Let(n, simplifyBin(defop(lhs), defop(rhs), op))
+
+      // Nothing
+      case _ => inst
+    }
+
+    simplifyExt(singleSimp, defop)
+  }
+
+  def simplifyBin(lhsDef: Option[Op], rhsDef: Option[Op], op: Op.Bin): Op = {
+    (lhsDef, rhsDef, op) match {
+
+      // (a * b) % a = 0
+      // (a * b) % b = 0
+      case (Some(Op.Bin(Imul, _, a, b)),
+            _,
+            Op.Bin(Srem, ty, _, c))
+          if ((a == c) || (b == c)) =>
+        Op.Copy(IVal(0, ty))
+
+      // (x << a) * b = x * (2^a * b)
+      case (Some(Op.Bin(Shl, _, x, IVal(a))),
+            _,
+            Op.Bin(Imul, ty, _, IVal(b))) =>
+        Op.Bin(Imul, ty, x, IVal(math.pow(2, a).toLong * b, ty))
+
+      // (x * a) << b = x * (a * 2^b)
+      case (Some(Op.Bin(Imul, _, x, IVal(a))),
+            _,
+            Op.Bin(Shl, ty, _, IVal(b))) =>
+        Op.Bin(Imul, ty, x, IVal(a * math.pow(2, b).toLong, ty))
+
+      // a + (0 - b) = a - b
+      case (_, Some(Op.Bin(Isub, _, IVal(0), b)), Op.Bin(Iadd, ty, a, _)) =>
+        Op.Bin(Isub, ty, a, b)
+
+      // a - (0 - b) = a + b
+      case (_, Some(Op.Bin(Isub, _, IVal(0), b)), Op.Bin(Isub, ty, a, _)) =>
+        Op.Bin(Iadd, ty, a, b)
+
+      // (0 - a) + b = b - a
+      case (Some(Op.Bin(Isub, _, IVal(0), a)), _, Op.Bin(Iadd, ty, _, b)) =>
+        Op.Bin(Isub, ty, b, a)
+
+      // (a - b) + b = a
+      case (Some(Op.Bin(Isub, _, a, b)), _, Op.Bin(Iadd, _, _, c))
+          if (b == c) =>
+        Op.Copy(a)
+
+      // (a + b) - b = a
+      case (Some(Op.Bin(Iadd, _, a, b)), _, Op.Bin(Isub, _, _, c))
+          if (b == c) =>
+        Op.Copy(a)
+
+      case _ => op
+    }
+  }
+
+  def simplifyExt(inst: Inst, defop: DefOp)(implicit fresh: Fresh): Seq[Inst] = {
+    inst match {
+      // (a * c) + (b * c) = (a + b) * c
+      case Let(n,
+               Op.Bin(Iadd,
+                      ty,
+                      defop(Op.Bin(Imul, _, a, c)),
+                      defop(Op.Bin(Imul, _, b, d)))) if (c == d) =>
+        val tmp = fresh()
+        Seq(Let(tmp, Op.Bin(Iadd, ty, a, b)),
+            Let(n, Op.Bin(Imul, ty, Val.Local(tmp, ty), c)))
+
+      // unbox (box a) = a
+      case Let(n, Op.Unbox(_, defop(Op.Box(_, a)))) =>
+        Seq(Let(n, Op.Copy(a)))
+
+      // box (unbox b) = b
```

Comment:
This one is not true. It could throw if `b` is not of the right type. And if `b eq null`, this should create a non-null box. In general, this pair of operation is not optimizable.